File: Source/BuildingSecurity.WebApp/ViewModels/AlarmManager/Alarm.js
/*----------------------------------------------------------------------------

(C) Copyright 2012 Johnson Controls, Inc.
Use or Copying of all or any part of this program, except as
permitted by License Agreement, is prohibited.

------------------------------------------------------------------------------*/

/*global ko: false, $: false, jci: false, HistoryEntry: false */

//
// This is the Alarm view model. It is the view model for each row in the active alarm list.
//
function Alarm(serializedAlarm, settings) {
    "use strict";

    var self = this;

    // Private methods
    function alarmHistoryRecieved(historyEntries) {
        var entry, deleteCount = 1, takeIndex = 0, index;

        // If the list of entries is empty do nothing.
        if (historyEntries.length === 0) {
            return;
        }

        // Only merge if it's likely to help performance
        if (self.historyEntries().length > 15 ||
                self.historyEntries().length > (historyEntries.length / 2)) {

            // How many entries have the latest timestamp
            entry = self.historyEntries()[0];
            while (deleteCount < self.historyEntries().length &&
                    entry.sortableTimestamp === self.historyEntries()[deleteCount].sortableTimestamp) {
                deleteCount += 1;
            }

            // Skip all the entries until we get to the latest timestamp
            historyEntries.sort(function (a, b) { return a.sortableTimestamp.localeCompare(b.sortableTimestamp); });
            while (takeIndex < historyEntries.length &&
                    entry.sortableTimestamp.localeCompare(historyEntries[takeIndex].sortableTimestamp) > 0) {
                takeIndex += 1;
            }
            historyEntries = historyEntries.slice(takeIndex);

            // Merge the new items if there are any
            if (deleteCount !== historyEntries.length) {
                self.historyEntries.splice(0, deleteCount);
                while (historyEntries.length >= 10) {
                    self.historyEntries.unshift(new HistoryEntry(historyEntries[9]), new HistoryEntry(historyEntries[8]),
                        new HistoryEntry(historyEntries[7]), new HistoryEntry(historyEntries[6]), new HistoryEntry(historyEntries[5]),
                        new HistoryEntry(historyEntries[4]), new HistoryEntry(historyEntries[3]), new HistoryEntry(historyEntries[2]),
                        new HistoryEntry(historyEntries[1]), new HistoryEntry(historyEntries[0]));
                    historyEntries.splice(0, 10);
                }
                for (index = 0; index < historyEntries.length; index += 1) {
                    self.historyEntries.unshift(new HistoryEntry(historyEntries[index]));
                }
            }
        } else {
            // Otherwise just update with updated list
            for (index = 0; index < historyEntries.length; index += 1) {
                historyEntries[index] = new HistoryEntry(historyEntries[index]);
            }

            self.historyEntries(historyEntries.sort(function (a, b) {
                return b.sortableTimestamp.localeCompare(a.sortableTimestamp);
            }));
        }
    }

    function updateHistory() {
        jci.buildingSecurityClient.getAlarmHistory(self.id, alarmHistoryRecieved);
    }

    // Data
    // Bound to the id of the details collapsible panel, also the id needed for calls to manipulate this alarm
    self.id = serializedAlarm.Id;

    // composite id's used for various elements displaying the alarm
    self.detailsId = self.id + "-details";
    self.detailsIdSelector = jci.makeIdSelector(self.detailsId);
    self.responseTextId = self.id + "-responseText";
    self.responseTextIdSelector = jci.makeIdSelector(self.responseTextId);
    self.enterResponseButtonId = self.id + "-enterResponseButton";
    self.enterResponseButtonIdSelector = jci.makeIdSelector(self.enterResponseButtonId);

    // Bound to the history list
    self.historyEntries = ko.observableArray();

    // Bound to the selected checkbox
    self.selected = ko.observable(false);

    // The date/time the alarm occurred - bound to a column - this is not observable - only set on creation, then static
    self.messageDateString = ko.observable(jci.spaceConsumingString(serializedAlarm.MessageDateString));
    self.messageTimeString = ko.observable(jci.spaceConsumingString(serializedAlarm.MessageTimeString));
    self.messageDateTimeString = ko.observable(jci.spaceConsumingString(serializedAlarm.MessageDateTimeString));
    self.messageUtcDateTime = ko.observable(serializedAlarm.MessageUtcDateTime);
    self.messageDisplayDateTimeString = ko.computed({
        read: function () {
            var dayOfAlarm;
            dayOfAlarm = new Date(self.messageUtcDateTime());
            if ((jci.currentObservableDate().getTime() - dayOfAlarm.getTime()) > (23 * 60 * 60 * 1000)) {//if the alarm is over 23hrs old
                return self.messageDateString();
            }
            return self.messageTimeString();
        },
        deferEvaluation: true
    });

    // Not bound, just used to sort based on occurred date time
    self.messageDateTimeSortableString = ko.observable(serializedAlarm.MessageDateTimeSortableString);

    // The priority of the alarm - bound to a column - this is not observable - only set on creation, then static
    self.priority = jci.spaceConsumingString(serializedAlarm.Priority);

    // The priority color of the alarm - bound to a style - this should be observable (though right now it is static) - should change with user preferences updates
    self.priorityColor = ko.observable(jci.alarmDisplayOptions.getPriorityColor(self.priority));
    jci.alarmDisplayOptions.updateCount.subscribe(function () {
        self.priorityColor(jci.alarmDisplayOptions.getPriorityColor(self.priority));
    });

    // The status of the alarm - bound to a column - this is observable and changes as the alarm is managed
    self.alarmStateDescription = ko.observable(jci.spaceConsumingString(serializedAlarm.AlarmStateDescription));

    // Display the alarm state description in the alarm item header.
    self.alarmStateDescriptionDispaly = ko.computed(function () {
        return " (" + this.alarmStateDescription() + ")";
    }, self);

    // The Source(description) of the alarm - bound to a column - this is observable and may change as the alarm is managed by the P2000
    self.description = ko.observable(jci.spaceConsumingString(serializedAlarm.Description));

    // The date/time of the last action taken on the alarm - bound to a column - this is observable and changes as the alarm is managed
    self.stateDateString = ko.observable(jci.spaceConsumingString(serializedAlarm.StateDateString));
    self.stateTimeString = ko.observable(jci.spaceConsumingString(serializedAlarm.StateTimeString));
    self.stateDateTimeString = ko.observable(jci.spaceConsumingString(serializedAlarm.StateDateTimeString));

    // Not bound, just used to sort based on action date timeThe date/time of the last action taken on the alarm
    self.stateDateTimeSortableString = ko.observable(serializedAlarm.StateDateTimeSortableString);

    // Incremented every time the triggering item changes state
    self.conditionSequence = ko.observable(serializedAlarm.ConditionSequence);

    // The site of the alarm - bound to the alarm details - this is not observable - only set on creation, then static
    self.site = jci.spaceConsumingString(serializedAlarm.Site);

    // Is the alarm public - bound to the alarm details - this is not observable - only set on creation, then static
    self.isPublicDescription = jci.spaceConsumingString(serializedAlarm.IsPublicDescription);

    // The category of the alarm - bound to the alarm details - this is not observable - only set on creation, then static
    self.category = jci.spaceConsumingString(serializedAlarm.Category);

    // The description of the current state of the alarm - bound to the alarm details - this is observable and changes as the alarm is managed
    self.sourceState = ko.observable(serializedAlarm.SourceState);
    self.sourceStateDescription = ko.observable(jci.spaceConsumingString(serializedAlarm.SourceStateDescription));

    self.isAlarmStateNA = ko.computed(function () {
        return self.sourceState() === -1;
    });

    self.isAlarmStateSecure = ko.computed(function () {
        return self.sourceState() === 0;
    });

    self.isAlarmStateAlarm = ko.computed(function () {
        return self.sourceState() === 1;
    });

    self.isAlarmStateShort = ko.computed(function () {
        return self.sourceState() === 2;
    });

    self.isAlarmStateOpen = ko.computed(function () {
        return self.sourceState() === 3;
    });

    self.isAlarmStateSuppressed = ko.computed(function () {
        return self.sourceState() === 4;
    });

    self.isAlarmStateUnknown = ko.computed(function () {
        return (!self.isAlarmStateNA() &&
            !self.isAlarmStateSecure() &&
                !self.isAlarmStateAlarm() &&
                    !self.isAlarmStateShort() &&
                        !self.isAlarmStateOpen() &&
                            !self.isAlarmStateSuppressed());
    });

    // The escalation value of the alarm - bound to the alarm details - this is observable and changes as the alarm is managed
    self.escalation = ko.observable(jci.spaceConsumingString(serializedAlarm.Escalation));

    // The instructions for the alarm - bound to the alarm details - this is not observable - only set on creation, then static
    self.instructions = serializedAlarm.Instructions;

    // Bound to the response text box in the alarm details
    self.enteredResponse = ko.observable(jci.string.empty());
    self.enteredResponse.subscribe(function (newValue) {
        self.isValidResponse(jci.string.trim(newValue) !== jci.string.empty());
    });

    self.isValidResponse = ko.observable(false);

    // Holds the error that may result from operations taken on the alarm
    self.error = ko.observable(jci.string.empty());

    // Computed value used to control visibility of the error block
    self.isErrorAvailable = ko.computed(function () {
        return (self.error() !== jci.string.empty());
    });

    self.isInstructionAvailable = ko.computed(function () {
        return (self.instructions !== jci.string.empty());
    });

    self.isEscalationVisible = ko.computed(function () {
        return ((+(self.escalation())) > 0);
    });

    // True if the Alarm is in a Pending state, otherwise false
    self.isPending = ko.observable(serializedAlarm.IsPending);

    // True if the Alarm can be Completed, otherwise false
    self.isCompletable = ko.observable(serializedAlarm.IsCompletable);

    // True if one can Respond to the Alarm, otherwise false
    self.isRespondable = ko.observable(serializedAlarm.IsRespondable);

    // True if the Alarm can be Removed, otherwise false
    self.isRemovable = ko.observable(serializedAlarm.IsRemovable);

    // True if the Alarm can be completed, otherwise false
    self.isCompleted = ko.observable(serializedAlarm.IsCompleted);

    self.isAcknowledgeVisible = ko.computed(function () {
        return self.isPending();
    });

    self.isCompleteVisible = ko.computed(function () {
        return !self.isAcknowledgeVisible() && !self.isCompleted();
    });

    self.isRemoveVisible = ko.computed(function () {
        return !self.isCompleteVisible() && !self.isAcknowledgeVisible();
    });

    // -1=Don't care, 0=Collapsed, 1=Expanded
    self.isExpandedValue = ko.observable(settings.autoExpand ? -1 : 0);

    // used to indicate whether we are expanded, bound to a css class - this is observable and changes as the alarm is managed
    self.isExpanded = ko.computed({
        read: function () {
            if (self.isExpandedValue() === 1) {
                return true;
            }

            if (self.isExpandedValue() === 0) {
                return false;
            }

            // To expand by default when State = Alarm: || (self.sourceState() === 1)
            return (self.isPending());
        },
        write: function (newValue) {
            self.isExpandedValue(newValue ? 1 : 0);
        }
    });

    if (self.isExpanded()) {
        updateHistory();
    }

    self.isExpanded.subscribe(function (newValue) {
        if (newValue) {
            updateHistory();
        }
    });

    // Return value of the specified sortKey formatted as a sortable type
    self.getSortableValue = function (sortKey) {
        var value;

        switch (sortKey) {
        case jci.alarmManagerColumns.Pending:
            return self.isPending() ? 1 : 0;
        case jci.alarmManagerColumns.OccurredDateTime:
            return self.messageDateTimeSortableString();
        case jci.alarmManagerColumns.Priority:
            return +self.priority;
        case jci.alarmManagerColumns.Status:
            return self.alarmStateDescription();
        case jci.alarmManagerColumns.Description:
            return self.description.toLowerCase();
        case jci.alarmManagerColumns.ActionDateTime:
            return self.stateDateTimeSortableString();
        case jci.alarmManagerColumns.Importance:
            // TODO: Need to support an arbitrary list of sort-able values
            value = 0;

            value += self.isPending() ? 10 : 0;
            value += self.isAlarmStateAlarm() ? 1 : 0;

            return value;
        }

        return null;
    };

    self.isAlarmCollapsed = ko.computed(function () {
        return self.isCompleteVisible() && !(self.isCompletable());
    });

    // Handle the details being shown
    self.detailsShown = function () {
        // Works in all browsers with double call - doesn't with a single
        $(self.responseTextIdSelector).focus().focus();
    };

    // Determine the background color depending on whether the alarm is pending.
    self.priorityColorBackground = ko.computed(function () {
        return this.isPending() ? this.priorityColor() : 'none';
    }, self);
    
    // Determine the highest pending status and alarm state.
    self.isHighStatusState = ko.computed(function () {
        return this.isAlarmStateAlarm() && this.isPending();
    }, self);

    // Determine if the current priority level is 'highest' and return appropriate visibility style value.
    self.setHighStatusStateVisible = ko.computed(function () {
        return this.isHighStatusState() === true ? "visible" : "hidden";
    }, self);

    // Determine the background color depending on whether the alarm is pending.
    self.highStatusStateColorBackground = ko.computed(function () {
        return this.isHighStatusState() ? this.priorityColor() : 'none';
    }, self);
    
    // Helpers - are not data bound to anything

    self.actionParams = function () {
        return { 'alarmId': self.id, 'conditionSequence': self.conditionSequence() };
    };

    self.update = function (alarmUpdates) {
        var propertyName;
        for (propertyName in alarmUpdates) {
            if (alarmUpdates.hasOwnProperty(propertyName) && ko.isObservable(self[propertyName])) {
                self[propertyName](alarmUpdates[propertyName]);
            }
        }
    };

    self.processChanges = function (alarmChanges) {
        self.alarmStateDescription(jci.spaceConsumingString(alarmChanges.AlarmStateDescription));
        self.description(jci.spaceConsumingString(alarmChanges.Description));
        self.messageDateString(jci.spaceConsumingString(alarmChanges.MessageDateString));
        self.messageTimeString(jci.spaceConsumingString(alarmChanges.MessageTimeString));
        self.messageDateTimeString(jci.spaceConsumingString(alarmChanges.MessageDateTimeString));
        self.messageUtcDateTime(alarmChanges.MessageUtcDateTime);
        self.messageDateTimeSortableString(alarmChanges.MessageDateTimeSortableString);
        self.stateDateString(jci.spaceConsumingString(alarmChanges.StateDateString));
        self.stateTimeString(jci.spaceConsumingString(alarmChanges.StateTimeString));
        self.stateDateTimeString(jci.spaceConsumingString(alarmChanges.StateDateTimeString));
        self.stateDateTimeSortableString(alarmChanges.StateDateTimeSortableString);
        self.sourceState(alarmChanges.SourceState);
        self.sourceStateDescription(jci.spaceConsumingString(alarmChanges.SourceStateDescription));
        self.escalation(jci.spaceConsumingString(alarmChanges.Escalation));
        self.conditionSequence(alarmChanges.ConditionSequence);
        self.isPending(alarmChanges.IsPending);
        self.isCompletable(alarmChanges.IsCompletable);
        self.isRespondable(alarmChanges.IsRespondable);
        self.isRemovable(alarmChanges.IsRemovable);
        self.isCompleted(alarmChanges.IsCompleted);

        if (self.isExpanded()) {
            updateHistory();
        }
    };
}

File: Source/JohnsonControls.BuildingSecurity/Entities/Alarm.cs
/*----------------------------------------------------------------------------

  (C) Copyright 2012 Johnson Controls, Inc.
      Use or Copying of all or any part of this program, except as
      permitted by License Agreement, is prohibited.

------------------------------------------------------------------------------*/

using System;
using System.Globalization;
using System.Runtime.Serialization;

namespace JohnsonControls.BuildingSecurity
{
    /// <summary>
    /// Represents an Alarm in P2000.  Inherits from Event
    /// </summary>
    [DataContract]
    public class Alarm : Message
    {
        /// <summary>
        /// Creates an immutable Alarm object
        /// </summary>
        /// <param name="id">Event id</param>
        /// <param name="description">Description of the Alarm</param>
        /// <param name="partitionId">The Id of the partition the event occurred on</param>
        /// <param name="partitionName">The name of the partition the event occurred on</param>
        /// <param name="isPublic">True if the event is public, otherwise false</param>
        /// <param name="messageDateTime">The Date and Time the alarm occurred.</param>
        /// <param name="alarmTypeDescription">The name of the alarm type</param>
        /// <param name="category">The category of the alarm</param>
        /// <param name="priority">The alarms priority</param>
        /// <param name="isResponseRequired">True if the alarm requires a response, otherwise false</param>
        /// <param name="isAcknowledgeRequired">True if the alarm requires acknowledgement, otherwise false</param>
        /// <param name="alarmState">The ID of the AlarmState or condition</param>
        /// <param name="alarmStateDescription">The name of the AlarmState or condition</param>
        /// <param name="stateDateTime">The timestamp of the last time the AlarmState change</param>
        /// <param name="conditionSequence">incremented every time the triggering item changes state</param>
        /// <param name="site">The site object from which the alarm was generated.</param>
        /// <param name="sourceState">The ID of the source state.</param>
        /// <param name="sourceStateDescription">The source state description.</param>
        /// <param name="escalation">The escalation level of the alarm.</param>
        /// <param name="instructions">The instructions given with this alarm.</param>
        /// <param name="isPublicDescription">The text description for the IsPublic field.</param>
        /// <param name="isPending">True if the Alarm is in a Pending state, otherwise false</param>
        /// <param name="isCompletable">True if the Alarm can be Completed, otherwise false</param>
        /// <param name="isRespondable">True if one can Respond to the Alarm, otherwise false</param>
        /// <param name="isRemovable">True if the Alarm can be Removed, otherwise false</param>
        /// <param name="isCompleted">True if the Alarm can be completed, otherwise false</param>
        public Alarm(Guid id, string description, Guid partitionId, string partitionName, bool isPublic, DateTimeOffset messageDateTime,
            string alarmTypeDescription, string category, int priority, bool isResponseRequired, bool isAcknowledgeRequired,
            int alarmState, string alarmStateDescription, DateTimeOffset stateDateTime, int conditionSequence, string site,
            int sourceState, string sourceStateDescription, int escalation, string instructions, string isPublicDescription,
            bool isPending,
            bool isCompletable,
            bool isRespondable,
            bool isRemovable,
            bool isCompleted
            )
            : base(id, description, partitionId, partitionName, isPublic, messageDateTime)
        {
            AlarmState = alarmState;
            AlarmStateDescription = alarmStateDescription;
            StateDateTime = stateDateTime;
            StateDateString = StateDateTime.ToString("d", CultureInfo.CurrentCulture);
            StateTimeString = StateDateTime.ToString("t", CultureInfo.CurrentCulture);
            StateDateTimeString = StateDateTime.ToString("G", CultureInfo.CurrentCulture);
            StateDateTimeSortableString = StateDateTime.ToString("s", CultureInfo.InvariantCulture);
            ConditionSequence = conditionSequence;

            IsAcknowledgeRequired = isAcknowledgeRequired;
            IsResponseRequired = isResponseRequired;
            Priority = priority;
            Category = category;
            AlarmTypeDescription = alarmTypeDescription;
            SourceState = sourceState;
            SourceStateDescription = sourceStateDescription;
            Site = site;
            Escalation = escalation;
            Instructions = instructions;
            IsPublicDescription = isPublicDescription;

            IsPending = isPending;
            IsCompletable = isCompletable;
            IsRespondable = isRespondable;
            IsRemovable = isRemovable;
            IsCompleted = isCompleted;
        }

        /// <summary>
        /// Gets the description of the alarm type.
        /// </summary>
        /// <value>
        /// The type of the alarm. example: "Muster Zone Status"
        /// </value>
        [DataMember]
        public string AlarmTypeDescription { get; private set; }

        /// <summary>
        /// Gets the category of the alarm.
        /// </summary>
        [DataMember]
        public string Category { get; private set; }

        /// <summary>
        /// Gets the alarm priority.
        /// </summary>
        /// <value>
        /// The alarm priority. (0 -255)
        /// </value>
        [DataMember]
        public int Priority { get; private set; }

        /// <summary>
        /// True if the alarm requires response before completion, otherwise false
        /// </summary>
        public bool IsResponseRequired { get; private set; }

        /// <summary>
        /// True if the alarm requires acknowledgement, otherwise false
        /// </summary>
        public bool IsAcknowledgeRequired { get; private set; }

        /// <summary>
        /// Gets the ID of the AlarmState
        /// </summary>
        public int AlarmState { get; private set; }

        /// <summary>
        /// Gets the label of the AlarmState
        /// </summary>
        /// <value>
        /// completed, responding, acknowledged, pending
        /// </value>
        [DataMember]
        public string AlarmStateDescription { get; private set; }

        /// <summary>
        /// The timestamp of the last time the AlarmState changed, represented in the local time.
        /// </summary>
        /// <remarks>
        /// This time is in the local time of the current machine.
        /// </remarks>
        public DateTimeOffset StateDateTime { get; private set; }

        /// <summary>
        /// Gets the date portion of the <see cref="StateDateTime"/> as a short date string formatted for the current culture
        /// </summary>
        [DataMember]
        public string StateDateString { get; private set; }

        /// <summary>
        /// Gets the time portion of the <see cref="StateDateTime"/> as a short time string formatted for the current culture
        /// </summary>
        [DataMember]
        public string StateTimeString { get; private set; }

        /// <summary>
        /// Returns the <see cref="StateDateTime"/> as a general date time (long time) string formatted for the current culture
        /// </summary>
        [DataMember]
        public string StateDateTimeString { get; private set; }

        /// <summary>
        /// Returns the <see cref="StateDateTime"/> as a sortable date time string
        /// </summary>
        [DataMember]
        public string StateDateTimeSortableString { get; private set; }

        /// <summary>
        /// Incremented every time the triggering item changes state
        /// </summary>
        [DataMember]
        public int ConditionSequence { get; private set; }

        /// <summary>
        /// Gets the source state.
        /// </summary>
        [DataMember]
        public int SourceState { get; private set; }

        /// <summary>
        /// Gets the source state description.
        /// </summary>
        [DataMember]
        public string SourceStateDescription { get; private set; }

        /// <summary>
        /// Gets the site name.
        /// </summary>
        [DataMember]
        public string Site { get; private set; }

        /// <summary>
        /// Gets the escalation level.
        /// </summary>
        /// <remarks>Valid value range is 0 - 255</remarks>
        [DataMember]
        public int Escalation { get; private set; }

        /// <summary>
        /// Gets the Instruction text for alarm
        /// </summary>
        [DataMember]
        public string Instructions { get; private set; }

        /// <summary>
        /// Gets the description concerning an alarm's public status
        /// </summary>
        [DataMember]
        public string IsPublicDescription { get; private set; }

        /// <summary>
        /// True if the Alarm is in a Pending state, otherwise false
        /// </summary>
        [DataMember]
        public bool IsPending { get; private set; }

        /// <summary>
        /// True if the Alarm can be Completed, otherwise false
        /// </summary>
        [DataMember]
        public bool IsCompletable { get; private set; }

        /// <summary>
        /// True if one can Respond to the Alarm, otherwise false
        /// </summary>
        [DataMember]
        public bool IsRespondable { get; private set; }

        /// <summary>
        /// True if the Alarm can be Removed, otherwise false
        /// </summary>
        [DataMember]
        public bool IsRemovable { get; private set; }

        /// <summary>
        /// True if the Alarm can be completed, otherwise false
        /// </summary>
        [DataMember]
        public bool IsCompleted { get; private set; }


        public Alarm CloneWith(Guid? id = null,
            string description = null,
            Guid? partitionId = null,
            string partitionName = null,
            bool? isPublic = null,
            DateTimeOffset? messageDateTime = null,
            string alarmTypeDescription = null,
            string category = null,
            int? priority = null,
            bool? isResponseRequired = null,
            bool? isAcknowledgeRequired = null,
            int? alarmState = null,
            string alarmStateDescription = null,
            DateTimeOffset? stateDateTime = null,
            int? conditionSequence = null,
            string site = null,
            int? sourceState = null,
            string sourceStateDescription = null,
            int? escalation = null,
            string instructions = null,
            string isPublicDescription = null,
            bool? isPending = null,
            bool? isCompletable = null,
            bool? isRespondable = null,
            bool? isRemovable = null,
            bool? isCompleted = null
            )
        {
            return new Alarm(id:id??Id,
                description:description??Description,
                partitionId:partitionId??PartitionId,
                partitionName:partitionName??PartitionName,
                isPublic:isPublic??IsPublic,
                messageDateTime:messageDateTime??MessageDateTime,
                alarmTypeDescription:alarmTypeDescription??AlarmTypeDescription,
                category:category??Category,
                priority:priority??Priority,
                isResponseRequired:isResponseRequired??IsResponseRequired,
                isAcknowledgeRequired:isAcknowledgeRequired??IsAcknowledgeRequired,
                alarmState:alarmState??AlarmState,
                alarmStateDescription: alarmStateDescription ?? AlarmStateDescription,
                stateDateTime:stateDateTime??StateDateTime,
                conditionSequence:conditionSequence??ConditionSequence,
                site:site??Site,
                sourceState:sourceState??SourceState,
                sourceStateDescription:sourceStateDescription??SourceStateDescription,
                escalation:escalation??Escalation,
                instructions:instructions??Instructions,
                isPublicDescription:isPublicDescription??IsPublicDescription,
                isPending:isPending??IsPending,
                isCompletable:isCompletable??IsCompletable,
                isRespondable:isRespondable??IsRespondable,
                isRemovable:isRemovable??IsRemovable,
                isCompleted:isCompleted??IsCompleted);
        }
    }
}